### DISCUSSION TITLE
How to write a recursive proof to use BN254 circuit to verify a BW6-761 proof

### DISCUSSION THREAD

**Discussion by matchv** [Consensys/gnark#887] [Q&A] [answered]
I've written my 1st circuit to verify BLS12-377 signatures with BW6-761 curve, it worked. I have the provingKey, verifyingKey, and public input of the 1st circuit.

But I got stuck writing the 2nd circuit using BN254 to verify the BW6 proof, there is no documentation and the playground example doesn't help, this is the core function

`if err := groth16_bn254.Verify(&circuit.InnerProof, &circuit.Vk, circuit.PublicInputs); err != nil {
		log.Panicf("Verify err: %s\n", err)
	}`

Params &circuit.InnerProof, &circuit.Vk, circuit.PublicInputs are the 1st circuits proof, and verification key converted to BN254 as below (the converted data):

// bw6_761 to bn254 data
`var bw6Json = `\{
  "InnerProof": {
    "Ar": {
      "X": "341245340993342336091611986059803560824819802620044319164102944257965656728",
      "Y": "13209611567594967319141461933660626474506744001703676332103399204349367314425"
    },
    "Krs": {
      "X": "18545994976034171503533418221235588381390441723613843006017391903670273719821",
      "Y": "12584722091595719049407959717805656538748102918472433370542704594009651263776"
    },
    "Bs": {
      "X": {
        "A0": "11143793221007836116646035823624337942039121669333804086516493513773421012438",
        "A1": "4720885926472301886935522344134254117821686941184619328203677875740508263479"
      },
      "Y": {
        "A0": "12149064360954419126082960309700899997625883051264047484140397186517453271821",
        "A1": "20692290263015728632103734515093104311170103134869341742186188293446065187070"
      }
    },
    "Commitments": [
      {
        "X": "14973059768597107327446759210879563591832803097465832166070691313070482771028",
        "Y": "6088817941494569712746427365416194310743238768765798571515085768289135653076"
      }
    ],
    "CommitmentPok": {
      "X": "10560913124437957273687821266309025900875273258209474501257360120153298431499",
      "Y": "9195747957480043011698251481835839943638971605095522684195280909268774504450"
    }
  },
  "Vk": {
    "G1": {
      "Alpha": {
        "X": "7570466509720200156140509749030428354083135043878430714311634123090277019334",
        "Y": "15668289559921995999647929069564682145127237571239907570067016262269553587586"
      },
      "Beta": {
        "X": "3128332467260591062006974521425999093794557056326279839971523182236035301731",
        "Y": "9101893981778563309185648330119618844127195199008409045991426353682989061652"
      },
      "Delta": {
        "X": "9450773229471261925921073577107218695879836008812642459493459318158828497843",
        "Y": "4853673499609593850650110838861328637990963901391066165404848465824967085115"
      },
      "K": [
        {
          "X": "12567687540875469737276177485342582592105821825199714287916515571919094601849",
          "Y": "16177692398511272801220273136611981149278528906746889774430354292511150225237"
        },
        {
          "X": "10530518020911126377074218334963714160878046561958067958104313982825542422592",
          "Y": "706474996070269270817396766416969319206108240635148592002338290969442304431"
        }
      ]
    },
    "G2": {
      "Beta": {
        "X": {
          "A0": "9300217224648796745325647142788721474465583182507830593498958209217462010800",
          "A1": "13439597396777995062927840774330638384726186434625853930072393213408079110013"
        },
        "Y": {
          "A0": "21857319329058477782953570866118265926617278745417878816622975639183642065053",
          "A1": "5049441465639544036653715487618573067133803159847450974161986822983127154796"
        }
      },
      "Delta": {
        "X": {
          "A0": "10179254693139657973324045851074838672031447379337713284350886364732280225505",
          "A1": "20315607521494548436407580795903212393843935598115863512259125078238995089987"
        },
        "Y": {
          "A0": "6499859875064411034861911380956008056653944956845947503227145215830492750943",
          "A1": "2022734318857292045182860555971547080898792113746739805556765370221348773170"
        }
      },
      "Gamma": {
        "X": {
          "A0": "9574648719632354189092059692454002898866729536065762720323422790227635883429",
          "A1": "7918058711233182988344472812674308602396461429401860220858276046043005280131"
        },
        "Y": {
          "A0": "19653517921949298637383552493188341360792553676860446663779750138173171789300",
          "A1": "17293405481506433424182185361852710915650104539067247861437903753201202533913"
        }
      }
    },
    "CommitmentKey": { },
    "PublicAndCommitmentCommitted": [
      [ ]
    ]
  },
  "PublicInputs": [ ]
}
`

The entire function to verify:

`func demo1() {
	var circuit Circuit
	if err := json.Unmarshal([]byte(bw6Json), &circuit); err != nil {
		log.Fatalf("unmarshal error: %v\n", err)
	}

	fmt.Printf("current CurveID: %s\n", circuit.Vk.CurveID())
	if err := groth16_bn254.Verify(&circuit.InnerProof, &circuit.Vk, circuit.PublicInputs); err != nil {
		log.Panicf("Verify err: %s\n", err)
	}
	fmt.Println(circuit)
}

func main() {
	demo1()
}
`

The error message:

`current CurveID: bn254
2023/10/23 09:04:50 Verify err: pairing doesn't match
panic: Verify err: pairing doesn't match
`
It looks I'm close to get it working but I may missed something or misunderstood something?

Also, this is to verify the 1st circuit's proof in a main() function, I tried to write the 2nd circuit for it but it gives me straight error, I can't import /backend/groth16/bn254 to do any in-circuit computation.

Can anybody write/explain to me a few lines of the correct in-circuit code or pseudo code, with the right import path, the right parameter set, the right data format?  I searched up and down but there is little documentation, thanks a lot.

**Comment by ivokub** [comment, marked as answer]
Hi, right now the in-circuit Groth16 verifier does not support verifying the random challenge we use for field emulation. There is some work before we can add it (using custom field to hash function in native Groth16 verifier and doing wrong-field mimc hashing in-circuit). We are right now working on it for PLONK and when we have the pieces it should be easy to add the support.

See https://github.com/Consensys/gnark/pull/880, https://github.com/Consensys/gnark/pull/879, https://github.com/Consensys/gnark/pull/884.
